field/src/array.rs

```diff
@@ -0,0 +1,147 @@
+use core::array;
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use crate::{
+    exp_10540996611094048183, exp_1717986917, exp_1725656503, exp_u64_by_squaring, AbstractField,
+    Field,
+};
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct FieldArray<F: Field, const N: usize>(pub [F; N]);
+
+impl<F: Field, const N: usize> Default for FieldArray<F, N> {
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+impl<F: Field, const N: usize> From<[F; N]> for FieldArray<F, N> {
+    fn from(arr: [F; N]) -> Self {
+        Self(arr)
+    }
+}
+
+impl<F: Field, const N: usize> AbstractField for FieldArray<F, N> {
+    const ZERO: Self = FieldArray([F::ZERO; N]);
+    const ONE: Self = FieldArray([F::ONE; N]);
+    const TWO: Self = FieldArray([F::TWO; N]);
+    const NEG_ONE: Self = FieldArray([F::NEG_ONE; N]);
+
+    fn from_bool(b: bool) -> Self {
+        [F::from_bool(b); N].into()
+    }
+
+    fn from_canonical_u8(n: u8) -> Self {
+        [F::from_canonical_u8(n); N].into()
+    }
+
+    fn from_canonical_u16(n: u16) -> Self {
+        [F::from_canonical_u16(n); N].into()
+    }
+
+    fn from_canonical_u32(n: u32) -> Self {
+        [F::from_canonical_u32(n); N].into()
+    }
+
+    fn from_canonical_u64(n: u64) -> Self {
+        [F::from_canonical_u64(n); N].into()
+    }
+
+    fn from_canonical_usize(n: usize) -> Self {
+        [F::from_canonical_usize(n); N].into()
+    }
+
+    fn from_wrapped_u32(n: u32) -> Self {
+        [F::from_wrapped_u32(n); N].into()
+    }
+
+    fn from_wrapped_u64(n: u64) -> Self {
+        [F::from_wrapped_u64(n); N].into()
+    }
+
+    fn multiplicative_group_generator() -> Self {
+        [F::multiplicative_group_generator(); N].into()
+    }
+
+    #[inline]
+    fn exp_u64(&self, power: u64) -> Self {
+        match power {
```

Comment:
This is a bit hacky... e.g. `1717986917` is only really useful for `Mersenne31`'s `x^{1/5}`, so semi-field-specific information is leaking into this more general struct.

I have a potential solution in mind, but it involves changing some core APIs, so might be best as a followup PR.

Reply:
I made #122 on top of this with the potential solution.